cbmc/proofs/proof_guide.md

```diff
@@ -0,0 +1,399 @@
+  ==>
+FORALL { int k; (0 <= k && k < len) ==> dst[k] == 0 }
+```
+
+`j <= len - 1` can be simplified to `j < len` so that simplifies to:
+
+```
+FORALL { int j; (0 <= j && j < len) ==> dst[j] == 0 }
+  ==>
+FORALL { int k; (0 <= k && k < len) ==> dst[k] == 0 }
+```
+
+which is True.
+
+## Recipe to prove a new function
+
+Let's say we want to develop a proof of a function. Here are the basic steps.
+1. Populate a proof directory
+2. Update Makefile
+3. Update harness function
+4. Supply top-level contracts for the function
+5. Supply loop-invariants (if requied) and other interior contracts
+6. Prove it!
+
+These steps are expanded on in the following sub-sections
+
+### Populate a proof directory
+
+For MLKEM-C-AArch64, proof directories lie below `cbmc/proofs`
+
+Create a new sub-directory in there, where the name of the directory is the name of the function _without_ the `$(MLKEM_NAMESPACE)` prefix.
+
+That directory needs to contain 3 files.
+
+* cbmc-proof.txt
+* Makefile
+* XXX_harness.c
+
+where "XXX" is the name of the function being proved - same as the directory name.
+
+I suggest that you copy these files from an existing proof directory and modify the latter two. The `cbmc-proof.txt` file is just a marker that says "this directory contains a CBMC proof" to the tools, so no modification is required.
+
+### Update Makefile
+
+The `Makefile` sets options and targets for this proof. Let's imagine that the function we want to prove is called `XXX` (without the `$(MLKEM_NAMESPACE)` prefix).
+
+Edit the Makefile and update the definition of the following variables:
+
+* HARNESS_FILE - should be `XXX_harness`
+* PROOF_UID - should be `XXX`
+* PROJECT_SOURCES - should the files containing the source code of XXX
+* CHECK_FUNCTION_CONTRACTS - set to the `XXX`, but _with_ the `$(MLKEM_NAMESPACE)` prefix if required
+* USE_FUNCTION_CONTRACTS - a list of functions that `XXX` calls where you want CBMC to use the contracts of the called function for proof, rather than 'inlining' the called function for proof. Include the `$(MLKEM_NAMESPACE)` prefix if required
+* CBMCFLAGS - additional flags to pass to the final run of CBMC. This is normally set to `--smt2` which tells cbmc to run Z3 as its underlying solver. Can also be set to `--bitwuzla` which is sometimes better at generaing counter-examples when Z3 fails.
+* FUNCTION_NAME - set to `XXX` with the `$(MLKEM_NAMESPACE)` prefix if required
+* CBMC_OBJECT_BITS. Normally set to 8, but might need to be increased if CBMC runs out of memory for this proof.
+
+For documentation of these (and the other) options, see the `cbmc/proofs/Makefile.common` file.
+
+### Update harness function
+
+The file `XXX_harness.c` should declare a single function called `XXX_harness()` that calls `XXX` exactly one, with appropriately typed parameters. Using contracts, this harness function should not need to contain any CBMC `ASSUME` or `ASSERT` statements at all.
```

Comment:
Can we expand on this? I get the impression that the harness MUST NOT use any stack allocated structures, but literally just declare variables of the right types and pass them to the function under proof, right? See #325

Reply:
See #340. @rod-chapman can you extend the guide accoringly?

Reply:
Done